### ISSUE TITLE
updating to 2.13.4 problem

### ISSUE THREAD

**Issue by svidimitrov** [Icinga/icinga2#9434]
## Describe the bug

The service crashes when updated to 2.13.4-1.el7

## Expected behavior

Updating to: 2.13.4-1.el7.icinga

## Your Environment

* Version used (`icinga2 --version`):  icinga2-2.13.3-1.el7.icinga.x86_64
* Operating System and version: Linux CentOS 7 
* Enabled features (`icinga2 feature list`): api checker ido-mysql mainlog
* Icinga Web 2 version and modules (System - About):
* Config validation (`icinga2 daemon -C`):
* If you run multiple Icinga 2 instances, the `zones.conf` file (or `icinga2 object list --type Endpoint` and `icinga2 object list --type Zone`) from all affected nodes.

## Additional context

Stacktrace:
0# icinga::Application::SigAbrtHandler(int) in /usr/lib64/icinga2/sbin/icinga2
1# 0x00007F59278E4630 in /lib64/libpthread.so.02.13.4-1.el7.icinga
2# gsignal in /lib64/libc.so.6
3# abort in /lib64/libc.so.6
4# 0x00007F592757FF67 in /lib64/libc.so.6
5# 0x00007F5927588329 in /lib64/libc.so.6
6# std::basic_string<char, std::char_traits, std::allocator >::~basic_string() in /lib64/libstdc++.so.6
7# icinga::ApplyRule::~ApplyRule() in /usr/lib64/icinga2/sbin/icinga2
8# void std::vector<icinga::ApplyRule, std:allocator Icinga::ApplyRule >::_M_realloc_insertIcinga::ApplyRule(__gnu_cxx::__normal_iterator<icinga::ApplyRule*, std::vector<icinga::ApplyRule, std::allocatoricinga::ApplyRule> > >, icinga::ApplyRule&&) in /usr/lib64/icinga2/sbin/icinga2
9# icinga::ApplyRule::AddRule(icinga::String const&, icinga::String const&, icinga::String const&, boost::intrusive_ptricinga::Expression> const&, boost::intrusive_ptr icinga::Expression> const&, icinga::String const&, icinga::String const&, icinga::String const&, boost::intrusive_ptricinga::Expression const&, bool, icinga::DebugInfo const&, boost::intrusive_ptricinga::Dictionary const&) in /usr/lib64/icinga2/sbin/icinga2
10# icinga::ApplyExpression::DoEvaluate(icinga::ScriptFrame&, icinga::DebugHint*) const in /usr/lib64/icinga2/sbin/icinga2
11# icinga::Expression::Evaluate(icinga::ScriptFrame&, icinga::DebugHint*) const in /usr/lib64/icinga2/sbin/icinga2
12# icinga::DictExpression::DoEvaluate(icinga::ScriptFrame&, icinga::DebugHint*) const in /usr/lib64/icinga2/sbin/icinga2
13# icinga::OwnedExpression::DoEvaluate(icinga::ScriptFrame&, icinga::DebugHint*) const in /usr/lib64/icinga2/sbin/icinga2
14# icinga::Expression::Evaluate(icinga::ScriptFrame&, icinga::DebugHint*) const in /usr/lib64/icinga2/sbin/icinga2
15# icinga::DictExpression::DoEvaluate(icinga::ScriptFrame&, icinga::DebugHint*) const in /usr/lib64/icinga2/sbin/icinga2
16# icinga::Expression::Evaluate(icinga::ScriptFrame&, icinga::DebugHint*) const in /usr/lib64/icinga2/sbin/icinga2
17# icinga::DictExpression::DoEvaluate(icinga::ScriptFrame&, icinga::DebugHint*) const in /usr/lib64/icinga2/sbin/icinga2
18# icinga::Expression::Evaluate(icinga::ScriptFrame&, icinga::DebugHint*) const in /usr/lib64/icinga2/sbin/icinga2
19# icinga::ConfigItem::Commit(bool) in /usr/lib64/icinga2/sbin/icinga2
20# 0x0000000000A5404C in /usr/lib64/icinga2/sbin/icinga2
21# icinga::WorkQueue::RunTaskFunction(std::function<void ()> const&) in /usr/lib64/icinga2/sbin/icinga2
22# 0x0000000000A49EDE in /usr/lib64/icinga2/sbin/icinga2
23# icinga::WorkQueue::RunTaskFunction(std::function<void ()> const&) in /usr/lib64/icinga2/sbin/icinga2
24# icinga::WorkQueue::WorkerThreadProc() in /usr/lib64/icinga2/sbin/icinga2
25# 0x00007F59297E4954 in /lib64/libboost_thread.so.1.69.0
26# 0x00007F59278DCEA5 in /lib64/libpthread.so.0
27# clone in /lib64/libc.so.6

**Comment by julianbrost** [comment]
Does this happen reliably, i.e. every time you start Icinga? Also, can you please try to figure out if something specific in your configuration is causing this?

**Comment by julianbrost** [comment]
Looks like there's missing synchronization in `ApplyRule::AddRule()`. This would explain why this happens. Nothing that was introduced in this release though, so I think this can also happen on older versions.

**Comment by julianbrost** [comment]
I just tried to reproduce this with no success, the code I suspected in https://github.com/Icinga/icinga2/issues/9434#issuecomment-1176305008 usually shouldn't run in parallel, so shouldn't need synchronization.

Do you happen to have nested `object`/`apply` directives like in https://github.com/Icinga/icinga2/issues/8633#issuecomment-790484532? When I do something like this, I can reproduce the crash, so I suspect this is a duplicate of that issue.

**Comment by julianbrost** [comment]
No feedback so far, but I'm quite certain that there's some nesting of `object` and `apply` given the stack trace: `icinga::ApplyExpression::DoEvaluate` (i.e. `apply`) within `icinga::Expression::Evaluate` called from `icinga::ConfigItem::Commit` (i.e. evaluating the body of some object).

Therefore closing this as a duplicate of #8633.
